bindings/serialize.cc

```diff
@@ -0,0 +1,405 @@
+  for (std::vector<Label>::iterator l = label.begin(); l != label.end(); ++l) {
+    encodeMessage(b, 3, &*l);
+  }
+}
+
+Profile::Profile(std::string periodType, std::string periodUnit, int64_t period,
+                 int64_t timeNanos, int64_t durationNanos,
+                 std::string dropFrames, std::string keepFramesX) {
+  // first index of strings must be ""
+  getStringId("");
+
+  this->periodType =
+      ValueType(getStringId(periodType), getStringId(periodUnit));
+  this->period = period;
+  this->timeNanos = timeNanos;
+  this->durationNanos = durationNanos;
+  this->dropFramesX = getStringId(dropFrames);
+  this->keepFramesX = getStringId(keepFramesX);
+  this->defaultSampleTypeX = 0;
+}
+
+void Profile::addSampleType(std::string type, std::string unit) {
+  int64_t typeX = getStringId(type);
+  int64_t unitX = getStringId(unit);
+  sampleType.push_back(ValueType(typeX, unitX));
+}
+
+uint64_t Profile::addSample(std::unique_ptr<Node> &node,
+                            std::deque<uint64_t> stack) {
+  uint64_t loc = getLocationId(node);
+  stack.push_front(loc);
+  std::vector<Sample> nodeSamples = node->getSamples(*this, stack);
+  sample.insert(sample.end(), nodeSamples.begin(), nodeSamples.end());
+  return loc;
+}
+
+uint64_t Profile::getLocationId(std::unique_ptr<Node> &node) {
+  std::string key = std::to_string(node->getFileID()) + ":" +
+                    std::to_string(node->getLineNumber()) + ":" +
+                    std::to_string(node->getColumnNumber()) + ":" +
+                    node->getName();
+  auto ids = functionIdMap.find(key);
+  if (ids != functionIdMap.end()) {
+    return ids->second;
+  }
+  uint64_t id = location.size() + 1;
+  std::vector<Line> lines;
+  lines.push_back(getLine(node));
+  ProfileLocation l = ProfileLocation(id, 0, 0, lines, false);
+  location.push_back(l);
+  return id;
+}
+
+Line Profile::getLine(std::unique_ptr<Node> &node) {
+  return Line(getFunctionId(node), node->getLineNumber());
+}
+
+int64_t Profile::getFunctionId(std::unique_ptr<Node> &node) {
+  std::string name = node->getName();
+  std::string key = std::to_string(node->getFileID()) + ":" + name;
+  auto ids = functionIdMap.find(key);
+  if (ids != functionIdMap.end()) {
+    return ids->second;
+  }
+  int64_t nameX = getStringId(name);
+  int64_t filenameX = getStringId(node->getFilename());
+  int64_t id = function.size() + 1;
+  ProfileFunction f =
+      ProfileFunction(id, nameX, nameX, filenameX, node->getLineNumber());
+  function.push_back(f);
+  return id;
+}
+
+int64_t Profile::getStringId(std::string s) {
+  auto pair = stringIdMap.find(s);
+  if (pair != stringIdMap.end()) {
+    return pair->second;
+  }
+  int64_t id = strings.size();
+  stringIdMap.insert(stringIdMap.begin(),
+                     std::pair<std::string, int64_t>(s, id));
+  strings.push_back(s);
+  return id;
+}
+
+void Profile::encode(std::vector<char> &b) {
+  for (std::vector<ValueType>::iterator x = sampleType.begin();
```

Comment:
For these iterations over arrays, frankly I think using index iteration would be more compact code:

```
for (size_t i = 0; i < sampleType.size(); i++) {
  // Use sampleType[i].
```

Reply:
Done.